File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarSleepUtils.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

private let ARABICA_USER_ROOT_FOLDER = "/U/0/"
private let SLEEP_DIRECTORY = "SLEEP/"
private let SLEEP_PROTO = "SLEEPRES.BPB"
private let NRST_DIRECTORY = "NSTRESUL/"
private let NRST_PROTO = "NSTRCONT.BPB"
private let dateFormat: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyyMMdd"
    formatter.locale = Locale(identifier: "en_US_POSIX")
    return formatter
}()
private let TAG = "PolarSleepUtils"

internal class PolarSleepUtils {

    static func readSleepFromDayDirectory(client: BlePsFtpClient, date: Date) async throws -> PolarSleepData.PolarSleepAnalysisResult {
        var result = try await readSleepData(client: client, date: date)
        result = try await readSleepSkinTemperatureResult(client: client, date: date, sleepAnalysisResult: result)
        return result
    }

    static func readSleepData(client: BlePsFtpClient, date: Date) async throws -> PolarSleepData.PolarSleepAnalysisResult {
        BleLogger.trace(TAG, "readSleepFromDayDirectory: \(date)")
        let sleepDataFilePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(SLEEP_DIRECTORY)\(SLEEP_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = sleepDataFilePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbSleepAnalysisResult(serializedBytes: Data(response))
            return PolarSleepData.PolarSleepAnalysisResult(
                sleepStartTime: try PolarTimeUtils.pbLocalDateTimeToDate(pbLocalDateTime: proto.sleepStartTime),
                sleepEndTime: try PolarTimeUtils.pbLocalDateTimeToDate(pbLocalDateTime: proto.sleepEndTime),
                lastModified: proto.hasLastModified ? try PolarTimeUtils.pbSystemDateTimeToDate(pbSystemDateTime: proto.lastModified) : nil,
                sleepGoalMinutes: proto.sleepGoalMinutes,
                sleepWakePhases: PolarSleepData.fromPbSleepwakePhasesListProto(pbSleepwakePhasesList: proto.sleepwakePhases),
                snoozeTime: try PolarSleepData.convertSnoozeTimeListToLocalTime(snoozeTimeList: proto.snoozeTime),
                alarmTime: proto.hasAlarmTime ? try PolarTimeUtils.pbLocalDateTimeToDate(pbLocalDateTime: proto.alarmTime) : nil,
                sleepStartOffsetSeconds: proto.sleepStartOffsetSeconds,
                sleepEndOffsetSeconds: proto.sleepStartOffsetSeconds,
                userSleepRating: proto.hasUserSleepRating ? PolarSleepData.SleepRating(rawValue: proto.userSleepRating.rawValue) : nil,
                deviceId: proto.hasRecordingDevice ? proto.recordingDevice.deviceID : nil,
                batteryRanOut: proto.hasBatteryRanOut ? proto.batteryRanOut : nil,
                sleepCycles: PolarSleepData.fromPbSleepCyclesList(pbSleepCyclesList: proto.sleepCycles),
                sleepResultDate: try PolarTimeUtils.pbDateToDateComponents(pbDate: proto.sleepResultDate),
                originalSleepRange: try PolarSleepData.fromPbOriginalSleepRange(pbOriginalSleepRange: proto.originalSleepRange)
            )
        } catch {
            BleLogger.trace("readSleepFromDayDirectory() failed for path: \(sleepDataFilePath), error: \(error). No sleep data?")
            return PolarSleepData.PolarSleepAnalysisResult(
                sleepStartTime: nil, sleepEndTime: nil, lastModified: nil, sleepGoalMinutes: nil,
                sleepWakePhases: nil, snoozeTime: nil, alarmTime: nil, sleepStartOffsetSeconds: nil,
                sleepEndOffsetSeconds: nil, userSleepRating: nil, deviceId: nil, batteryRanOut: nil,
                sleepCycles: nil, sleepResultDate: nil, originalSleepRange: nil
            )
        }
    }

    static func readSleepSkinTemperatureResult(client: BlePsFtpClient, date: Date, sleepAnalysisResult: PolarSleepData.PolarSleepAnalysisResult) async throws -> PolarSleepData.PolarSleepAnalysisResult {
        BleLogger.trace(TAG, "readSleepSkinTemperature: \(date)")
        var result = sleepAnalysisResult
        let filePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(NRST_DIRECTORY)\(NRST_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = filePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let proto = try Data_PbSleepSkinTemperatureResult(serializedBytes: Data(response))
            result.sleepSkinTemperatureResult = try PolarSleepData.fromPbSleepTemperatureResult(pbSleepTemperatureResult: proto)
        } catch {
            BleLogger.trace("readSleepSkinTemperature() failed for path: \(filePath), error: \(error). No sleep skin temperature data?")
        }
        return result
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarSleepUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.sleep.*
import com.polar.services.datamodels.protobuf.SleepSkinTemperatureResult
import fi.polar.remote.representation.protobuf.SleepanalysisResult
import protocol.PftpRequest
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

private const val ARABICA_USER_ROOT_FOLDER = "/U/0/"
private const val SLEEP_DIRECTORY = "SLEEP/"
private const val SLEEP_PROTO = "SLEEPRES.BPB"
private const val NRST_DIRECTORY = "NSTRESUL/"
private const val NRST_PROTO = "NSTRCONT.BPB"
private val dateFormatter = DateTimeFormatter.ofPattern("yyyyMMdd", Locale.ENGLISH)
private const val TAG = "PolarSleepUtils"

internal object PolarSleepUtils {

    /**
     * Read sleep data for a given date.
     */
    suspend fun readSleepDataFromDayDirectory(
        client: BlePsFtpClient,
        date: LocalDate
    ): PolarSleepAnalysisResult {
        val response = readSleepData(client, date)
        return readSleepSkinTemperatureResult(client, date, response)
    }

    /**
     * Read sleep data.
     */
    private suspend fun readSleepData(client: BlePsFtpClient, date: LocalDate): PolarSleepAnalysisResult {
        BleLogger.d(TAG, "readSleepData: $date")
        return try {
            val sleepFilePath = "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/${SLEEP_DIRECTORY}${SLEEP_PROTO}"
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(sleepFilePath)
                    .build()
                    .toByteArray()
            )
            val proto = SleepanalysisResult.PbSleepAnalysisResult.parseFrom(response.toByteArray())
            PolarSleepAnalysisResult(
                PolarTimeUtils.pbLocalDateTimeToZonedDateTime(proto.sleepStartTime),
                PolarTimeUtils.pbLocalDateTimeToZonedDateTime(proto.sleepEndTime),
                PolarTimeUtils.pbSystemDateTimeToZonedDateTime(proto.lastModified),
                proto.sleepGoalMinutes,
                fromPbSleepwakePhasesListProto(proto.sleepwakePhasesList),
                convertSnoozeTimeListToZonedDateTimeList(proto.snoozeTimeList),
                if (proto.hasAlarmTime()) {
                    PolarTimeUtils.pbLocalDateTimeToZonedDateTime(proto.alarmTime)
                } else null,
                proto.sleepStartOffsetSeconds ?: null,
                proto.sleepEndOffsetSeconds ?: null,
                if (proto.hasUserSleepRating()) {
                    SleepRating.from(proto.userSleepRating.number)
                } else null,
                proto.recordingDevice.deviceId ?: null,
                proto.batteryRanOut ?: null,
                fromPbSleepCyclesList(proto.sleepCyclesList),
                PolarTimeUtils.pbDateToLocalDate(proto.sleepResultDate) ?: null,
                if (proto.hasOriginalSleepRange()) {
                    fromPbOriginalSleepRange(proto.originalSleepRange)
                } else null,
                null
            )
        } catch (_: Throwable) {
            PolarSleepAnalysisResult(
                null, null, null, null,
                null, null, null, null,
                null, null, null, null,
                null, null, null, null
            )
        }
    }

    /**
     * Read skin temperature data.
     */
    private suspend fun readSleepSkinTemperatureResult(
        client: BlePsFtpClient,
        date: LocalDate,
        sleepAnalysisResult: PolarSleepAnalysisResult
    ): PolarSleepAnalysisResult {
        BleLogger.d(TAG, "readSleepSkinTemperatureResult: $date")
        return try {
            val result = sleepAnalysisResult
            val sleepFilePath = "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/${NRST_DIRECTORY}${NRST_PROTO}"
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(sleepFilePath)
                    .build()
                    .toByteArray()
            )
            val proto = SleepSkinTemperatureResult.PbSleepSkinTemperatureResult.parseFrom(response.toByteArray())
            if (proto.hasSleepDate()) {
                result.sleepSkinTemperatureResult = fromPbSleepSkinTemperatureResult(proto)
            }
            result
        } catch (_: Throwable) {
            sleepAnalysisResult
        }
    }
}